### PR TITLE
ccgx: add hp dmc dock to ccgx plugin

### PR DESCRIPTION
**Pull request by chlee75**
Add HP Dock to ccgx plugin
 - HP Dock use Cypress DMC (Dock Management Controller) architecture
 - In DMC architecture, a single composite firmware binary file includes multiple firmware images and DMC update all devices connected to DMC including DMC itself.
 - fu-ccgx-dmc-device/fu-ccgx-dmc-firmware/fu-ccgx-dmc-common are added newly for DMC device.
 - HP USB-C Dock G5 and HP USB-C/A Universal Dock G2 are added to ccgx.quirk.
 - DMC has two kinds of Update Model
    Current HP Dock use DMC_UPDATE_MODEL_DOWNLOAD_TIRGGER
    1. DMC_UPDATE_MODEL_DOWNLOAD_TIRGGER (internal flash model)
      - DMC store all firmware images to an internal flash in a dock.
    2. DMC_UPDATE_MODEL_PENDING_RESET 
      - DMC update each firmware to secondary image flash on each device 
 - DMC has two kinds of update options
    1.  Update now (default)
      - DMC updates all device firmware after it receives all firmware images.
    2. Update on disconnect (see has-manual-replug flag)
       - DMC updates all device firmware when a dock is unplugged manually by a user.

### CODE REVIEW

**Comment by chlee75** [comment]
Hi. Richard.
 - cab files have been uploaded at  [ https://github.com/hughsie/cypress-firmware/pull/4](url) 
 - fu_ccgx_hpi_write_firmware_image() in fu-ccgx-dmc-device.c is a typo and it will be replaced with fu_ccgx_dmc_write_firmware_image in next commit.

File: plugins/ccgx/fu-ccgx-dmc-common.c [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-ccgx-dmc-common.h"
+
+gchar *
+fu_ccgx_dmc_version_to_string (guint32 val)
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-common.c:12]: Isn't this just `fu_common_version_from_uint32(FWUPD_VERSION_FORMAT_QUAD)`?

File: plugins/ccgx/fu-ccgx-dmc-common.c [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+fu_ccgx_dmc_update_model_to_string (DmcUpdateModel val)
+{
+	if (val == DMC_UPDATE_MODEL_UNKNOWN)
+		return "Update Model  Unknown";
+	if (val == DMC_UPDATE_MODEL_DOWNLOAD_TIRGGER)
+		return "Update Model Download Trigger";
+	if (val == DMC_UPDATE_MODEL_PENDING_RESET)
+		return "Update Model Pending Reset";
+	return NULL;
+
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-common.c:31]: no newline here

File: plugins/ccgx/fu-ccgx-dmc-common.h [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#pragma once
+
+#include <glib.h>
+
+/*******************************************************************************
+ * Defines
+*******************************************************************************/
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-common.h:13]: delete this

File: plugins/ccgx/fu-ccgx-dmc-common.h [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+#define DMC_FWCT_LENGTH_OFFSET      		4
+#define DMC_FWCT_MAX_SIZE           		2048
+
+#define DMC_CUSTOM_META_LENGTH_FIELD_SIZE  	2
+#define DMC_CUSTOM_META_LENGTH_OFFSET      	0
+#define DMC_CUSTOM_META_MAX_SIZE           	256
+
+/*******************************************************************************
+ * Enums
+*******************************************************************************/
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-common.h:65]: delete

File: plugins/ccgx/fu-ccgx-dmc-common.h [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+ 	 * 2 : new structure */
+	guint8          structure_version;
+	guint8          cdtt_version;
+	guint16         vid;
+	guint16         pid;
+	guint16         device_id;
+	gchar           vendor_string[DMC_IDENTITY_STRING_LEN];
+	gchar           product_string[DMC_IDENTITY_STRING_LEN];
+	guint8          custom_meta_data_flag;
+	/* model feild indicates the type of the firmware upgrade status
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-common.h:209]: field

File: plugins/ccgx/fu-ccgx-dmc-common.h [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+	guint8         length;
+	guint8         data[DMC_INTERRUPT_DATA_LEN];
+} DmcIntRqt;
+
+/* this structure defines header structure of FWCT */
+typedef struct __attribute__((packed)) {
+	guint32         signature;
+	guint16         size;
+	guint8          checksum;
+	guint8          version;
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-common.h:274]: I think you want to align the fields with tabs, not spaces.

File: plugins/ccgx/fu-ccgx-dmc-common.h [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+typedef struct __attribute__((packed)) {
+	guint8          img_id;
+	guint8          type;
+	guint16         start_row;
+	guint16         num_rows; /* size */
+	guint8          reserv0[2];
+} FwctSegmentationInfo;
+
+gchar   *fu_ccgx_dmc_version_to_string (guint32 val);
+gchar   *fu_ccgx_dmc_update_model_to_string (DmcUpdateModel val);
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-common.h:310]: copy the prototype style in existing files please.

File: plugins/ccgx/fu-ccgx-dmc-device.c [hunk 8]
```diff
@@ -0,0 +1,703 @@
...
+#include "fu-ccgx-dmc-firmware.h"
+
+#define DMC_FW_WRITE_STATUS_RETRY_DELAY_MS 30 /* 30 msec */
+
+struct _FuCcgxDmcDevice {
+	FuUsbDevice		parent_instance;
+	FWImageType		fw_image_type;
+        DmcDockIdentity		dock_id;
+        guint8			ep_intr_in;
+        guint8			ep_bulk_out;
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-device.c:24]: tabs

File: plugins/ccgx/fu-ccgx-dmc-device.c [hunk 9]
```diff
@@ -0,0 +1,703 @@
...
+{
+	if (!g_usb_device_control_transfer (fu_usb_device_get_dev (FU_USB_DEVICE (self)),
+					    G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					    G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					    G_USB_DEVICE_RECIPIENT_DEVICE,
+					    DMC_RQT_CODE_SOFT_RESET, /* request */
+					    reset_later, /* value */
+					    0, /* index */
+					    0, /* data */
+                                            0,  /* length */
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-device.c:145]: tabs

File: plugins/ccgx/fu-ccgx-dmc-device.c [hunk 10]
```diff
@@ -0,0 +1,703 @@
...
+                                            NULL, /* actual length */
+					    DMC_CONTROL_TRANSFER_DEFAULT_TIMEOUT, /* timeout */
+					    NULL, error)) {
+		g_prefix_error (error, "send reset error:");
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-device.c:149]: you need a single space after the `:` -- there are a lot of these to fix :)

File: plugins/ccgx/fu-ccgx-dmc-device.c [hunk 11]
```diff
@@ -0,0 +1,703 @@
...
+		return FALSE;
+	}
+	return TRUE;
+}
+
+
+static gboolean
+fu_ccgx_dmc_device_send_start_upgrade (FuCcgxDmcDevice *self,
+				       guint8* custom_meta_data,
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-device.c:158]: the `*` on the name please

File: plugins/ccgx/fu-ccgx-dmc-device.c [hunk 12]
```diff
@@ -0,0 +1,703 @@
...
+				       guint16 custom_meta_size,
+				       GError **error)
+{
+	guint16 value = 0;
+	if (custom_meta_size > 0)
+		value = 1;
+	if (custom_meta_data == NULL && custom_meta_size > 0) {
+		g_set_error(error,
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-device.c:166]: space before bracket

File: plugins/ccgx/fu-ccgx-dmc-device.c [hunk 13]
```diff
@@ -0,0 +1,703 @@
...
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "invalid  pointer at %s",__func__);
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-device.c:169]: don't do this. if it's really a critical internal error that can't happen just use `g_return_val_if_fail()`

File: plugins/ccgx/fu-ccgx-dmc-device.c [hunk 14]
```diff
@@ -0,0 +1,703 @@
...
+			      guint8* fwct_data,
+			      guint16 fwct_size,
+			      GError **error)
+{
+	if (fwct_data == NULL) {
+		g_set_error(error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "invalid  pointer at %s",__func__);
+			return FALSE;
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-device.c:225]: funky aligning

File: plugins/ccgx/fu-ccgx-dmc-device.c [hunk 15]
```diff
@@ -0,0 +1,703 @@
...
+			row_buffer = g_bytes_get_data (data_rcd->data, &row_size);
+			/* write row data */
+			if (!fu_ccgx_dmc_device_send_row_data (self, row_buffer, (guint16)row_size, error))
+				return FALSE;
+			/* increase fw written size */
+			self->fw_data_written += row_size;
+			/* update update progress */
+			fu_device_set_progress_full (device, (gsize) self->fw_data_written, (gsize) self->fw_data_size);
+			/* get reponse  */
+			for (guint8 retry = 0; retry < 2; retry++) {
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-device.c:373]: I think you want fu_device_retry here.

File: plugins/ccgx/fu-ccgx-dmc-device.c [hunk 16]
```diff
@@ -0,0 +1,703 @@
...
+					g_set_error (error,
+						     FWUPD_ERROR,
+						     FWUPD_ERROR_NOT_SUPPORTED,
+						     "invalid dmc intr req opcode in fw write = %d",
+						     dmc_int_req.opcode);
+						return FALSE;
+				}
+				if (dmc_int_req.data[0] == 0) /* success */
+					break;
+				else if (dmc_int_req.data[0] != 1)  {/* retry if data[0] is 1 otherwise error */
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-device.c:395]: if we `break` then you don't need the `else`

File: plugins/ccgx/fu-ccgx-dmc-device.c [hunk 17]
```diff
@@ -0,0 +1,703 @@
...
+	g_debug ("fw write request with meta:%u fwct:%u", (guint32)custom_meta_size, (guint32)fwct_size);
+	if (!fu_ccgx_dmc_device_send_fwct (self,fwct_data, fwct_size, error))
+		return FALSE;
+
+
+	/* init fw written */
+	self->fw_data_written = 0;
+	/* get total fw size */
+	self->fw_data_size = fu_ccgx_dmc_firmware_get_fw_data_size (FU_CCGX_DMC_FIRMWARE(firmware));
+	g_debug ("updating fw...");
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-device.c:465]: not helpful...

File: plugins/ccgx/fu-ccgx-dmc-device.c [hunk 18]
```diff
@@ -0,0 +1,703 @@
...
+	}
+
+	/* send fwct data */
+	g_debug ("fw write request with meta:%u fwct:%u", (guint32)custom_meta_size, (guint32)fwct_size);
+	if (!fu_ccgx_dmc_device_send_fwct (self,fwct_data, fwct_size, error))
+		return FALSE;
+
+
+	/* init fw written */
+	self->fw_data_written = 0;
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-device.c:462]: no, `self` represents attributes about the device, not progress.

File: plugins/ccgx/fu-ccgx-dmc-device.c [hunk 19]
```diff
@@ -0,0 +1,703 @@
...
+	FuCcgxDmcDevice *self = FU_CCGX_DMC_DEVICE (device);
+	GBytes* custom_meta_record = NULL;
+	gboolean custom_meta_exist = FALSE;
+
+	/* parse all images */
+	if (!fu_firmware_parse (firmware, fw, flags, error))
+		return NULL;
+
+	/* get custom meta record */
+	custom_meta_record = fu_ccgx_dmc_firmware_get_custom_meta_record (FU_CCGX_DMC_FIRMWARE(firmware));
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-device.c:539]: space before bracket pls

File: plugins/ccgx/fu-ccgx-dmc-device.c [hunk 20]
```diff
@@ -0,0 +1,703 @@
...
+	if (custom_meta_record)
+		if (g_bytes_get_size (custom_meta_record) > 0)
+			custom_meta_exist = TRUE;
+
+	/* check custom meta flag */
+	if ( self->dock_id.custom_meta_data_flag  !=  custom_meta_exist)  {
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-device.c:545]: no space after `if (` and weird double spaces

File: plugins/ccgx/fu-ccgx-dmc-device.c [hunk 21]
```diff
@@ -0,0 +1,703 @@
...
+		g_debug ("pending reset:%u",(guint32)manual_replug);
+
+		if (!fu_ccgx_dmc_device_send_sort_reset (self,
+							       manual_replug,
+							       error)) {
+			g_prefix_error (error, "soft reset error:");
+			return FALSE;
+		}
+	}  else {
+			g_set_error (error,
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-device.c:593]: intent level?

File: plugins/ccgx/fu-ccgx-dmc-device.c [hunk 22]
```diff
@@ -0,0 +1,703 @@
...
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "invalid update model = %u",
+				     self->update_model);
+			return FALSE;
+	}
+
+	if (manual_replug)
+		return TRUE;
+
+	g_debug ("reset and wait replug");
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-device.c:604]: not helpful

File: plugins/ccgx/fu-ccgx-dmc-device.c [hunk 23]
```diff
@@ -0,0 +1,703 @@
...
+
+	remove_delay_ms = fu_device_get_install_duration(self) * 1000;
+	fu_device_set_remove_delay (FU_DEVICE (self), remove_delay_ms); /* over a minute to flash offline */
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-device.c:607]: is the replug delay really proportional to the amount of time taken to flash?

File: plugins/ccgx/fu-ccgx-dmc-device.c [hunk 24]
```diff
@@ -0,0 +1,703 @@
...
+	g_autofree gchar *version = NULL;
+	g_autofree gchar *instance_id = NULL;
+
+	/* get dock identity */
+	if (!fu_ccgx_dmc_device_get_dock_id (self, &dock_id, error))
+		return FALSE;
+
+	/* store dock identity */
+	if (!fu_memcpy_safe ((guint8 *)&self->dock_id, sizeof(DmcDockIdentity), 0x0, /* dst */
+		     (guint8*)&dock_id, sizeof(DmcDockIdentity), 0, sizeof(DmcDockIdentity), error)) /* src */
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-device.c:629]: weird intending here

File: plugins/ccgx/fu-ccgx-dmc-device.c [hunk 25]
```diff
@@ -0,0 +1,703 @@
...
+		return FALSE;
+
+	/* get dock status */
+	if (!fu_ccgx_dmc_device_get_dock_status (self, &dock_status, error))
+		return FALSE;
+
+	/* set instance id */
+	instance_id = g_strdup_printf ("USB\\VID_%04X&PID_%04X&DMC",
+					fu_usb_device_get_vid (FU_USB_DEVICE (self)),
+					fu_usb_device_get_pid (FU_USB_DEVICE (self)));
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-device.c:639]: Do you need this custom Instance ID? Why not just use the standard USB one?

File: plugins/ccgx/fu-ccgx-dmc-device.c [hunk 26]
```diff
@@ -0,0 +1,703 @@
...
+	version_raw = dock_status.composite_version;
+	version = fu_ccgx_dmc_version_to_string (version_raw);
+	fu_device_set_version (FU_DEVICE (self), version);
+	fu_device_set_version_raw (FU_DEVICE (self), version_raw);
+
+	g_debug  ("[dmc dock] VID:%04X, PID:%04X, model:%d, vendor:%s, product:%s",
+		dock_id.vid, dock_id.pid, dock_id.model, dock_id.vendor_string, dock_id.product_string);
+
+	g_debug  ("[dmc status] status:0x%X, device count:%d, version:%s",
+		dock_status.device_status, dock_status.device_count, version);
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-device.c:653]: not helpful; if you want this make sure it's in the `_to_string()` handler.

File: plugins/ccgx/fu-ccgx-dmc-firmware.c [hunk 27]
```diff
@@ -0,0 +1,332 @@
...
+			g_autoptr(FuCcgxDmcFirmwareDataRecord) data_rcd = NULL;
+			/* alloc data record */
+			data_rcd = g_new0 (FuCcgxDmcFirmwareDataRecord, 1);
+			/* set row data offset */
+			fw_buffer_offset = row_data_offset;
+			/* read row data */
+			if (!fu_memcpy_safe (row_buff, actual_row_size, 0x0, /* dst */
+			     fw_buff, fw_buff_size, fw_buffer_offset, actual_row_size, error)) /* src */
+				return FALSE;
+			/* update hash */
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-firmware.c:152]: this whole section is a bit "dense" -- perhaps some newlines there?

File: plugins/ccgx/fu-ccgx-dmc-firmware.c [hunk 28]
```diff
@@ -0,0 +1,332 @@
...
+		*segment_info_offset += sizeof(FwctSegmentationInfo);
+	}
+
+	/* check checksum */
+	g_checksum_get_digest (csum, hash, &digestlen);
+	if (memcmp (hash, img_rcd->info_header.img_digest, DMC_HASH_SIZE) != 0) {
+		g_set_error_literal (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "invalid hash");
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-firmware.c:173]: maybe add what was expected and obtained?

File: plugins/ccgx/fu-ccgx-dmc-firmware.c [hunk 29]
```diff
@@ -0,0 +1,332 @@
...
+static void
+fu_ccgx_dmc_firmware_finalize (GObject *object)
+{
+	FuCcgxDmcFirmware *self = FU_CCGX_DMC_FIRMWARE (object);
+	if (self->fwct_record)
+		g_bytes_unref (self->fwct_record);
+	if (self->custom_meta_record)
+		g_bytes_unref (self->custom_meta_record);
+	if (self->image_records)
+		g_ptr_array_unref (self->image_records);
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-firmware.c:314]: please use ` != NULL` when comparing pointers

File: plugins/ccgx/fu-ccgx-dmc-firmware.h [hunk 30]
```diff
@@ -0,0 +1,35 @@
...
+
+typedef struct {
+        FwctImageInfo           info_header;
+        GPtrArray               *seg_records;
+} FuCcgxDmcFirmwareImageRecord;
+
+FuFirmware	*fu_ccgx_dmc_firmware_new			(void);
+GPtrArray	*fu_ccgx_dmc_firmware_get_image_records		(FuCcgxDmcFirmware	*self);
+GBytes	        *fu_ccgx_dmc_firmware_get_fwct_record		(FuCcgxDmcFirmware	*self);
+GBytes	        *fu_ccgx_dmc_firmware_get_custom_meta_record	(FuCcgxDmcFirmware	*self);
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-dmc-firmware.h:33]: tabs?

File: plugins/ccgx/fu-plugin-ccgx.c [hunk 31]
```diff
@@ -12,12 +12,17 @@
...
+#include "fu-ccgx-dmc-device.h"
+#include "fu-ccgx-dmc-firmware.h"
+
 
 void
 fu_plugin_init (FuPlugin *plugin)
 {
 	fu_plugin_set_build_hash (plugin, FU_BUILD_HASH);
 	fu_plugin_add_firmware_gtype (plugin, "ccgx", FU_TYPE_CCGX_FIRMWARE);
+        fu_plugin_add_firmware_gtype (plugin, "ccgx", FU_TYPE_CCGX_DMC_FIRMWARE);
```
Comment (hughsie) [inline plugins/ccgx/fu-plugin-ccgx.c:24]: tab

File: plugins/ccgx/fu-ccgx-dmc-common.c [hunk 32]
```diff
@@ -0,0 +1,32 @@
...
+fu_ccgx_dmc_version_to_string (guint32 val)
+{
+	return g_strdup_printf ("%u.%u.%u.%u",
+				(val >> 24) & 0xff,
+				(val >> 16) & 0xff,
+				(val >> 8) & 0xff,
+				(val) & 0xff);
+}
+
+gchar *
```
Comment (superm1) [inline plugins/ccgx/fu-ccgx-dmc-common.c:21]: shouldn't this be const?

File: plugins/ccgx/fu-ccgx-dmc-common.c [hunk 33]
```diff
@@ -0,0 +1,32 @@
...
+fu_ccgx_dmc_update_model_to_string (DmcUpdateModel val)
+{
+	if (val == DMC_UPDATE_MODEL_UNKNOWN)
+		return "Update Model  Unknown";
```
Comment (superm1) [inline plugins/ccgx/fu-ccgx-dmc-common.c:25]: extra spaces

File: plugins/ccgx/fu-ccgx-dmc-common.c [hunk 34]
```diff
@@ -0,0 +1,32 @@
...
+	if (val == DMC_UPDATE_MODEL_DOWNLOAD_TIRGGER)
```
Comment (superm1) [inline plugins/ccgx/fu-ccgx-dmc-common.c:26]: spelled trigger wrong

File: plugins/ccgx/fu-ccgx-dmc-common.c [hunk 35]
```diff
@@ -0,0 +1,32 @@
...
+{
+	return g_strdup_printf ("%u.%u.%u.%u",
+				(val >> 24) & 0xff,
+				(val >> 16) & 0xff,
+				(val >> 8) & 0xff,
+				(val) & 0xff);
+}
+
+gchar *
+fu_ccgx_dmc_update_model_to_string (DmcUpdateModel val)
```
Comment (superm1) [inline plugins/ccgx/fu-ccgx-dmc-common.c:22]: I would argue that all of these strings don't need the prefix "Update Model".

File: plugins/ccgx/fu-ccgx-dmc-device.c [hunk 36]
```diff
@@ -0,0 +1,703 @@
...
+fu_ccgx_dmc_device_get_dock_id (FuCcgxDmcDevice *self,
+			        DmcDockIdentity* dock_id,
+			        GError **error)
+{
+	if (dock_id == NULL) {
+		g_set_error(error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "invalid  pointer at %s",__func__);
+			return FALSE;
```
Comment (superm1) [inline plugins/ccgx/fu-ccgx-dmc-device.c:42]: Is this better to use `g_return_val_if_fail` ?

File: plugins/ccgx/fu-ccgx-dmc-common.c (hunk 2 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-common.c:19]: Add new line.

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 12 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-device.c:157]: it is removed.

File: plugins/ccgx/fu-ccgx-dmc-firmware.c (hunk 27 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-firmware.c:153]: Added new line

File: plugins/ccgx/fu-ccgx-dmc-common.c (hunk 1 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-common.c]: Replaced it with fu_common_version_from_uint32 on commit today.

File: plugins/ccgx/fu-ccgx-dmc-common.h (hunk 3 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-common.h]: Removed it on today commit

File: plugins/ccgx/fu-ccgx-dmc-common.h (hunk 4 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-common.h]: Removed it on today commit

File: plugins/ccgx/fu-ccgx-dmc-common.c (hunk 2 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-common.c:19]: Added new line on commit today (ccgx: fix code alignment, modify code for review).

File: plugins/ccgx/fu-ccgx-dmc-common.h (hunk 5 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-common.h]: Fixed it on commit today (ccgx: fix code alignment, modify code for review).

File: plugins/ccgx/fu-ccgx-dmc-common.h (hunk 6 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-common.h]: Fixed it on commit today (ccgx: fix code alignment, modify code for review).

File: plugins/ccgx/fu-ccgx-dmc-common.h (hunk 7 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-common.h]: Replace it with fu_ccgx_dmc_update_model_type_to_string() on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 8 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-device.c]: Fixed it on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 9 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-device.c]: Fixed it on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 10 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-device.c]: Fixed it on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 11 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-device.c]: Fixed it on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 13 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-device.c]: Replace it with g_return_val_if_fail() on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 14 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-device.c]: Removed this code on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 15 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-device.c]: Apply fu_device_retry on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 16 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-device.c]: Fixed it on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 17 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-device.c]: Removed it on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 18 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-device.c]: Replace it with local variable on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 19 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-device.c]: Fixed it on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 20 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-device.c]: Fixed it on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 21 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-device.c]: Fixed it on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 22 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-device.c]: Removed it on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 23 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-device.c]: Add InstallDuration and RemoveDelay on ccgx.quirk file on today commit (ccgx: fix code alignment, modify code for review).
Could you check whether RemoveDelay is right usage?

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 24 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-device.c:593]: It memory copy for structure variable.
Is there any good way for this?

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 25 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-device.c]: I removed DMC and use standard USB on today commit (ccgx: fix code alignment, modify code for review).

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 26 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-device.c]: Removed on today commit (ccgx: fix code alignment, modify code for review).
Could you let me know how to see _to_sting() information?

File: plugins/ccgx/fu-ccgx-dmc-firmware.c (hunk 28 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-firmware.c]: HASH size is 32 byte. How to print 32 byte value?

File: plugins/ccgx/fu-ccgx-dmc-firmware.c (hunk 29 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-firmware.c]: Fixed on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-ccgx-dmc-firmware.h (hunk 30 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-firmware.h]: Fixed on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-plugin-ccgx.c (hunk 31 above)
Comment (chlee75) [reply plugins/ccgx/fu-plugin-ccgx.c]: Fixed on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-ccgx-dmc-common.c (hunk 32 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-common.c]: add const on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-ccgx-dmc-common.c (hunk 33 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-common.c]: Fixed on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-ccgx-dmc-common.c (hunk 34 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-common.c]: Fixed on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 36 above)
Comment (chlee75) [reply plugins/ccgx/fu-ccgx-dmc-device.c]: Fixed on today commit (ccgx: fix code alignment, modify code for review)

File: plugins/ccgx/fu-ccgx-dmc-device.c [hunk 37]
```diff
@@ -0,0 +1,676 @@
...
+		self->fw_image_type = fu_ccgx_fw_image_type_from_string (value);
+		if (self->fw_image_type != FW_IMAGE_TYPE_UNKNOWN)
+			return TRUE;
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "invalid ImageKind");
+		return FALSE;
+	}
+	if (g_strcmp0 (key, "RemoveDelay") == 0) {
```
Comment (superm1) [inline plugins/ccgx/fu-ccgx-dmc-device.c:634]: @hughsie I feel like this is the sort of thing that we want in proper daemon quirks, not this custom quirk and custom code that ferries the attribute around only to call `fu_device_set_remove_delay`

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 26 above)
Comment (hughsie) [reply plugins/ccgx/fu-ccgx-dmc-device.c]: any time the daemon calls fu_device_to_string() -- so run the daemon with --verbose.

File: plugins/ccgx/fu-ccgx-dmc-device.c (hunk 37 above)
Comment (hughsie) [reply plugins/ccgx/fu-ccgx-dmc-device.c:634]: Added to master, good idea.

**Comment by hughsie** [comment]
I've fixed up a lot of small nits and rebased it against master. It works great on the DMC device I have here.

The CI failure is:

    Unknown tag changelog-should-mention-nmu at /usr/share/perl5/Lintian/Profile.pm line 389.

...but that's nothing to do with this PR.

File: plugins/ccgx/fu-ccgx-dmc-firmware.c (hunk 28 above)
Comment (hughsie) [reply plugins/ccgx/fu-ccgx-dmc-firmware.c]: So maybe not super useful. We can add it later if we hit this.

**Pull request merged**